apps/site/lib/site_web/controllers/stop_controller.ex

```diff
@@ -192,19 +192,44 @@ defmodule SiteWeb.StopController do
     json(conn, route_patterns_by_route_and_headsign(stop_id))
   end
 
-  @type by_route_and_headsign :: %{Route.id_t() => %{String.t() => [RoutePattern.t()]}}
+  @type headsign_info :: %{
+          required(:direction_id) => 0 | 1,
+          required(:route_patterns) => [RoutePattern.t()]
+        }
+  @type by_route_and_headsign :: %{Route.id_t() => %{String.t() => headsign_info}}
   @spec route_patterns_by_route_and_headsign(Stop.id_t()) :: by_route_and_headsign()
   defp route_patterns_by_route_and_headsign(stop_id) do
     stop_id
     |> RoutePatterns.Repo.by_stop_id()
-    |> Enum.reject(&not_serving_today?/1)
-    |> Enum.reject(&ends_at?(&1, stop_id))
-    |> Enum.map(&add_polyline/1)
-    |> Enum.group_by(& &1.route_id)
-    |> Enum.map(fn {k, v} -> {k, Enum.group_by(v, & &1.headsign)} end)
+    |> Stream.reject(&ends_at?(&1, stop_id))
+    |> Stream.chunk_by(& &1.route_id)
+    |> Stream.map(fn [%RoutePattern{route_id: route_id} | _] = route_route_patterns ->
+      by_headsign =
+        route_route_patterns
+        |> Stream.chunk_by(& &1.headsign)
+        |> Stream.map(fn [%RoutePattern{headsign: headsign} | _] = headsign_route_patterns ->
+          {headsign, with_headsign_info(headsign_route_patterns)}
+        end)
+        |> Map.new()
+
+      {route_id, by_headsign}
+    end)
     |> Map.new()
   end
 
+  @spec with_headsign_info([RoutePattern.t()]) :: headsign_info
```

Comment:
suggestion: Consider renaming this - `with_headsign_info` reads to me like it is only adding headsign information to the route patterns, but this is also filtering & adding polylines